Scripts/Runtime/Entities/Lifecycle/AbstractEntityLifecycleStatusSystem.cs

```diff
@@ -0,0 +1,303 @@
+{
+    /// <summary>
+    /// Provides information on the lifecycle of entities.
+    /// </summary>
+    /// <remarks>
+    /// Best practice is to extend this system and place the newly created system at a point in time when structural
+    /// changes have been made. Ex. Right after a <see cref="EntityCommandBufferSystem"/>.
+    /// Created or Imported entities will be treated as "Arrivals" to the World.
+    /// Destroyed or Evicted entities will be treated as "Departures" from the World.
+    /// Entities that are in a "CleanUp" state can be treated as "Departures" before they are actually destroyed by
+    /// Unity.
+    /// </remarks>
+    [AlwaysUpdateSystem]
+    public abstract partial class AbstractEntityLifecycleStatusSystem : AbstractAnvilSystemBase
+    {
+        private readonly List<EntityLifecycleStatus> m_EntityLifecycleStatus;
+        private readonly bool m_ShouldCleanupEntitiesCountAsDepartures;
+        private WorldEntityState m_WorldEntityState;
+        private NativeArray<JobHandle> m_Dependencies;
+        
+        /// <summary>
+        /// Creates a new <see cref="AbstractEntityLifecycleStatusSystem"/>
+        /// </summary>
+        /// <param name="shouldCleanupEntitiesCountAsDepartures">
+        /// If true, Entities that are in a CleanUp state with <see cref="ISystemStateComponentData"/>
+        /// will be considered as Departures. They will not show up later when they are actually destroyed by Unity.
+        /// </param>
+        protected AbstractEntityLifecycleStatusSystem(bool shouldCleanupEntitiesCountAsDepartures)
+        {
+            m_ShouldCleanupEntitiesCountAsDepartures = shouldCleanupEntitiesCountAsDepartures;
+            m_EntityLifecycleStatus = new List<EntityLifecycleStatus>();
+        }
+
+        protected override void OnCreate()
+        {
+            base.OnCreate();
+
+            m_WorldEntityState = WorldEntityState.GetOrCreate(World, this);
+
+            foreach (EntityLifecycleStatus entityLifecycleStatus in m_EntityLifecycleStatus)
+            {
+                entityLifecycleStatus.CreateQuery();
+            }
+
+            m_Dependencies = new NativeArray<JobHandle>(m_EntityLifecycleStatus.Count, Allocator.Persistent);
+        }
+
+        protected override void OnDestroy()
+        {
+            m_EntityLifecycleStatus.DisposeAllAndTryClear();
+            m_Dependencies.Dispose();
+
+            m_WorldEntityState.RemoveLifecycleStatusSystem(this);
+
+            base.OnDestroy();
+        }
+        
+        /// <summary>
+        /// Creates a new <see cref="IEntityLifecycleStatus"/> object to monitor Arrivals/Departures from a certain
+        /// archetype.
+        /// </summary>
+        /// <param name="componentTypes">The components to filter on.</param>
+        /// <returns>The <see cref="IEntityLifecycleStatus"/> object</returns>
+        protected IEntityLifecycleStatus CreateEntityLifecycleStatus(params ComponentType[] componentTypes)
```

Comment:
In actual usage does this ever get called outside of OnCreate?

I'm wondering if we should make an abstract method that the subclass has to implement and pass back all the different queries it wants to track. Knowing to call this is easy to overlook.

Alternatively, could we just pass all the different component queries that we want back in the constructor? They don't require World, EntityManager or the like to create.

Reply:
This function is called by your derived class to create an object that listens/reports the lifecycle status for a certain set of components.

```csharp
        public RenderingEntityLifecycleStatusSystem() : base(true)
        {
            EnvironmentRendererLifecycleStatus = CreateEntityLifecycleStatus(
                ComponentType.ReadOnly<EnvironmentRendererTag>(),
                ComponentType.ReadOnly<EnvironmentID>());
    
            ActorRendererLifecycleStatus = CreateEntityLifecycleStatus(
                ComponentType.ReadOnly<ActorRendererTag>(),
                ComponentType.ReadOnly<ActorID>());
        }
```

I don't think you'll overlook it because you need it to create the object anyway. 

Abstract method would be called by the constructor so I think it's probably better to just leave as the constructor.

I think you're thinking you'd extend this for each set of components you want to track, but it's more that you'd extend the system for each structural point in a frame you want to measure at and then create the inner lifecycle status objects for each set of components you want to track at that point.

Reply:
Ah ok, I didn't realize you'd keep reference to the created status objects. I'd just let consumers create this object and then register it with the subclass rather than having a factory method but this is ok for now.